### PR TITLE
Rm const traits in libcore

### PR DESCRIPTION
**Pull request by fee1-dead**
See [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/.60const.20Trait.60.20removal.20or.20rework) and https://github.com/rust-lang/rust/issues/110395

* [x] Bless ui tests
* [ ] Re constify some unstable functions with workarounds if they are needed

### CODE REVIEW

**Comment by fee1-dead** [comment]
cc @rust-lang/wg-const-eval

File: library/core/src/lib.rs [hunk 1]
```diff
@@ -207,9 +198,9 @@
 #![feature(const_mut_refs)]
 #![feature(const_precise_live_drops)]
 #![feature(const_refs_to_cell)]
+#![feature(const_trait_impl)]
```
Comment (oli-obk) [inline library/core/src/lib.rs:201]: why is this still needed?

Comment (oli-obk) [reply library/core/src/lib.rs:201]: edit: ah nevermind, Destruct is still const_trait, but not used anywhere like that in libcore

**Comment by the8472** [comment]
If `#[const_trait]` is getting removed shouldn't the `#[rustc_do_not_const_check]` get ripped out too?

**Comment by fee1-dead** [comment]
> If `#[const_trait]` is getting removed shouldn't the `#[rustc_do_not_const_check]` get ripped out too?

Do you mean those on the `Iterator` trait? They would be readded anyways, once we have a nice compiler implementation so is there a benefit in removing them now? I'd like to hear libs input on this.

**Comment by the8472** [comment]
Ok, if they're expected to be still applicable in the followup solution then they can stay. It wasn't clear to me how much "back to the drawing board" removal of the const-trait mechanism implied.

**Comment by oli-obk** [comment]
@bora retry network issue

**Comment by fee1-dead** [comment]
seems.. spurious? Local run on aarch64 says that it is working fine. I will try using [these instructions](https://rustc-dev-guide.rust-lang.org/tests/ci.html#using-ci-to-test).

**Comment by fee1-dead** [comment]
:/ Let's hope that readding the const stuff back will fix the tests.

@bors r=oli-obk

**Comment by Thomasdezeeuw** [comment]
Would have been nice to link to this in the tracking issue (https://github.com/rust-lang/rust/issues/92391) or even mention in the pr that the `const_cmp` was being removed. This was quite difficult to find (using GitHub search).

**Comment by alexschrod** [comment]
Just out of curiosity, since this has basically disabled everything that was done in #91930, should `#![feature(const_option_ext)]` be deprecated/removed? For the time being, `#![feature(const_option_ext)]` does nothing anymore, as all those methods are back to being non-`const`. I'm sure there are also features for other types (like `Result`) that have caused similar "pulled the `const` rug out from under them" consequences.

File: tests/ui/const-generics/generic_const_exprs/unify-op-with-fn-call.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+// known-bug: #110395
```
Comment (jackh726) [inline tests/ui/const-generics/generic_const_exprs/unify-op-with-fn-call.rs:1]: sorry for the necro:

This is a bad annotation. This test no longer tests what it was intended to. Also, #110395 isn't a bug...

**Pull request merged**